ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2837,264 @@ public sealed class Bot : IAsyncDisposable {
+			// Level 5 is maximum level for card badges according to https://steamcommunity.com/tradingcards/faq
+			linkElements.AddRange(badgePage.SelectNodes("//div[contains(@class, 'badge_row') and .//div[@class='badge_info_description']/div[contains(text(), 'Level 5')]]/a[@class='badge_row_overlay']"));
+
+			if (linkElements.Count == 0) {
+				return new HashSet<uint>(0);
+			}
+
+			HashSet<uint> result = new HashSet<uint>(linkElements.Count);
+
+			foreach (string? badgeUri in linkElements.Select(htmlNode => htmlNode.GetAttribute("href"))) {
+				if (string.IsNullOrEmpty(badgeUri)) {
+					ArchiLogger.LogNullError(nameof(badgeUri));
+
+					return null;
+				}
+
+				// URIs to foil badges are the same as for normal badges except they end with "?border=1"
+				string? appIDText = badgeUri.Split('/', StringSplitOptions.RemoveEmptyEntries)[^1].Split('?', StringSplitOptions.RemoveEmptyEntries)[0];
+
+				if (!uint.TryParse(appIDText, out uint appID) || (appID == 0)) {
+					ArchiLogger.LogNullError(nameof(appID));
+
+					return null;
+				}
+
+				result.Add(appID);
+			}
+
+			return result;
+		}
+
+		private bool CompleteSetLootScheduled;
+
+		private async Task SendCompletedSets() {
+			lock (SendCompleteSetsSemaphore) {
+				if (CompleteSetLootScheduled) {
+					return;
+				}
+
+				CompleteSetLootScheduled = true;
+			}
+
+			await SendCompleteSetsSemaphore.WaitAsync().ConfigureAwait(false);
+
+			try {
+				lock (SendCompleteSetsSemaphore) {
+					CompleteSetLootScheduled = false;
+				}
+
+				HashSet<uint>? appIDs = await GetPossiblyCompletedBadgeAppIDs().ConfigureAwait(false);
+
+				if ((appIDs == null) || (appIDs.Count == 0)) {
+					return;
+				}
+
+				HashSet<Steam.Asset> inventory;
+
+				try {
+					inventory = await ArchiWebHandler.GetInventoryAsync(SteamID)
+						.Where(item => item.Tradable)
+						.Where(item => (item.Type == Steam.Asset.EType.TradingCard) || (item.Type == Steam.Asset.EType.FoilTradingCard))
+						.Where(item => appIDs.Contains(item.RealAppID))
+						.ToHashSetAsync()
+						.ConfigureAwait(false);
+				} catch (HttpRequestException e) {
+					ArchiLogger.LogGenericWarningException(e);
+
+					return;
+				} catch (Exception e) {
+					ArchiLogger.LogGenericException(e);
+
+					return;
+				}
+
+				if (inventory.Count == 0) {
+					ArchiLogger.LogGenericWarning(string.Format(Strings.ErrorIsEmpty), nameof(inventory));
+
+					return;
+				}
+
+				Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), List<uint>> inventorySets = Trading.GetInventorySets(inventory);
+				List<Task<((uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity) Key, (uint Sets, byte CardsPerBadge) Value)>> loadCardCountsTasks = inventorySets.Select(async kv => (kv.Key, (Sets: kv.Value.FirstOrDefault(), CardsPerBadge: await ArchiWebHandler.GetCardCountForGame(kv.Key.RealAppID).ConfigureAwait(false)))).ToList();
```

Comment:
All tasks are launched here irrelevant of `OptimizationMode`. I'm also not sure if you need `ToList()` at all, but if you do, you probably can use `ToHashSet()` as well.

Reply:
Also regarding `await ArchiWebHandler.GetCardCountForGame` - is it possible to load card count while the inventory is loading? If not, we can select all the appIDs and load card counts for them in `Utilities.InParallel` in order to get rid of async lambda here.